File: tests/machine/x/pascal/group_by_having.pas
program GroupByHaving;
{$mode objfpc}
{$modeswitch nestedprocvars}

uses SysUtils, fgl, fphttpclient, Classes, Variants, fpjson, jsonparser;

type
  generic TArray<T> = array of T;

  generic _Group<T> = record
    Key: Variant;
    Items: specialize TArray<T>;
  end;

  generic function _group_by<T>(src: specialize TArray<T>; keyfn: function(it: T): Variant):
                                                                                          specialize
                                                                                             TArray<
                                                                                          specialize
                                                                                             _Group<
                                                                                             T>>;

var i,j,idx: Integer;
  key: Variant;
  ks: string;
begin
  SetLength(Result, 0);
  for i := 0 to High(src) do
    begin
      key := keyfn(src[i]);
      ks := VarToStr(key);
      idx := -1;
      for j := 0 to High(Result) do
        if VarToStr(Result[j].Key) = ks then
          begin
            idx := j;
            Break;
          end;
      if idx = -1 then
        begin
          idx := Length(Result);
          SetLength(Result, idx + 1);
          Result[idx].Key := key;
          SetLength(Result[idx].Items, 0);
        end;
      SetLength(Result[idx].Items, Length(Result[idx].Items)+1);
      Result[idx].Items[High(Result[idx].Items)] := src[i];
    end;
end;

generic procedure _json<T>(v: T);
begin
  writeln('[]');
end;

var
  _tmp0: specialize TFPGMap<string, string>;
  _tmp1: specialize TFPGMap<string, string>;
  _tmp10: specialize TFPGMap<string, Variant>;
  _tmp2: specialize TFPGMap<string, string>;
  _tmp3: specialize TFPGMap<string, string>;
  _tmp4: specialize TFPGMap<string, string>;
  _tmp5: specialize TFPGMap<string, string>;
  _tmp6: specialize TFPGMap<string, string>;
  _tmp7: specialize TArray<specialize TFPGMap<string, string>>;
  _tmp8: specialize TArray<specialize _Group<specialize TFPGMap<string, string>>>;
  _tmp9: specialize TArray<specialize TFPGMap<string, Variant>>;
  big: specialize TArray<specialize TFPGMap<string, Variant>>;
  p: specialize TFPGMap<string, string>;
  people: specialize TArray<specialize TFPGMap<string, string>>;

begin
  _tmp0 := specialize TFPGMap<string, string>.Create;
  _tmp0.AddOrSetData('name', 'Alice');
  _tmp0.AddOrSetData('city', 'Paris');
  _tmp1 := specialize TFPGMap<string, string>.Create;
  _tmp1.AddOrSetData('name', 'Bob');
  _tmp1.AddOrSetData('city', 'Hanoi');
  _tmp2 := specialize TFPGMap<string, string>.Create;
  _tmp2.AddOrSetData('name', 'Charlie');
  _tmp2.AddOrSetData('city', 'Paris');
  _tmp3 := specialize TFPGMap<string, string>.Create;
  _tmp3.AddOrSetData('name', 'Diana');
  _tmp3.AddOrSetData('city', 'Hanoi');
  _tmp4 := specialize TFPGMap<string, string>.Create;
  _tmp4.AddOrSetData('name', 'Eve');
  _tmp4.AddOrSetData('city', 'Paris');
  _tmp5 := specialize TFPGMap<string, string>.Create;
  _tmp5.AddOrSetData('name', 'Frank');
  _tmp5.AddOrSetData('city', 'Hanoi');
  _tmp6 := specialize TFPGMap<string, string>.Create;
  _tmp6.AddOrSetData('name', 'George');
  _tmp6.AddOrSetData('city', 'Paris');
  people := specialize TArray<specialize TFPGMap<string, string>>([_tmp0, _tmp1, _tmp2, _tmp3, _tmp4
            , _tmp5, _tmp6]);
  SetLength(_tmp7, 0);
  for p in people do
    begin
      _tmp7 := Concat(_tmp7, [p]);
    end;
  _tmp8 := specialize _group_by<specialize TFPGMap<string, string>>(_tmp7, function(p: specialize
           TFPGMap<string, string>): Variant begin Result := p.KeyData['city']
end
);
SetLength(_tmp9, 0);
for g in _tmp8 do
  begin
    _tmp10 := specialize TFPGMap<string, Variant>.Create;
    _tmp10.AddOrSetData('city', g.key);
    _tmp10.AddOrSetData('num', Length(g));
    _tmp9 := Concat(_tmp9, [_tmp10]);
  end;
big := _tmp9;
specialize _json<specialize TArray<specialize TFPGMap<string, Variant>>>(big);
end.

File: tests/machine/x/cpp/group_by_having.cpp
#include <iostream>
#include <map>
#include <string>
#include <unordered_map>
#include <vector>

template <typename T> void __json(const T &);
inline void __json(int v) { std::cout << v; }
inline void __json(double v) { std::cout << v; }
inline void __json(bool v) { std::cout << (v ? "true" : "false"); }
inline void __json(const std::string &v) { std::cout << "\"" << v << "\""; }
inline void __json(const char *v) { std::cout << "\"" << v << "\""; }
template <typename T> void __json(const std::vector<T> &v) {
  std::cout << "[";
  bool first = true;
  for (const auto &x : v) {
    if (!first)
      std::cout << ",";
    first = false;
    __json(x);
  }
  std::cout << "]";
}
template <typename K, typename V> void __json(const std::map<K, V> &m) {
  std::cout << "{";
  bool first = true;
  for (const auto &kv : m) {
    if (!first)
      std::cout << ",";
    first = false;
    __json(kv.first);
    std::cout << ":";
    __json(kv.second);
  }
  std::cout << "}";
}
template <typename K, typename V>
void __json(const std::unordered_map<K, V> &m) {
  std::cout << "{";
  bool first = true;
  for (const auto &kv : m) {
    if (!first)
      std::cout << ",";
    first = false;
    __json(kv.first);
    std::cout << ":";
    __json(kv.second);
  }
  std::cout << "}";
}

struct People {
  decltype(std::string("Alice")) name;
  decltype(std::string("Paris")) city;
};
inline bool operator==(const People &a, const People &b) {
  return a.name == b.name && a.city == b.city;
}
inline bool operator!=(const People &a, const People &b) { return !(a == b); }
struct __struct2 {
  decltype(std::declval<People>().city) key;
  std::vector<People> items;
};
inline bool operator==(const __struct2 &a, const __struct2 &b) {
  return a.key == b.key && a.items == b.items;
}
inline bool operator!=(const __struct2 &a, const __struct2 &b) {
  return !(a == b);
}
struct Big {
  decltype(std::declval<__struct2>().key) city;
  int num;
};
inline bool operator==(const Big &a, const Big &b) {
  return a.city == b.city && a.num == b.num;
}
inline bool operator!=(const Big &a, const Big &b) { return !(a == b); }
inline void __json(const Big &v) {
  bool first = true;
  std::cout << "{";
  if (!first)
    std::cout << ",";
  first = false;
  std::cout << "\"city\":";
  __json(v.city);
  if (!first)
    std::cout << ",";
  first = false;
  std::cout << "\"num\":";
  __json(v.num);
  std::cout << "}";
}
inline void __json(const People &v) {
  bool first = true;
  std::cout << "{";
  if (!first)
    std::cout << ",";
  first = false;
  std::cout << "\"name\":";
  __json(v.name);
  if (!first)
    std::cout << ",";
  first = false;
  std::cout << "\"city\":";
  __json(v.city);
  std::cout << "}";
}
int main() {
  std::vector<People> people =
      std::vector<People>{People{std::string("Alice"), std::string("Paris")},
                          People{std::string("Bob"), std::string("Hanoi")},
                          People{std::string("Charlie"), std::string("Paris")},
                          People{std::string("Diana"), std::string("Hanoi")},
                          People{std::string("Eve"), std::string("Paris")},
                          People{std::string("Frank"), std::string("Hanoi")},
                          People{std::string("George"), std::string("Paris")}};
  auto big = ([&]() {
    std::vector<__struct2> __groups;
    for (auto p : people) {
      auto __key = p.city;
      bool __found = false;
      for (auto &__g : __groups) {
        if (__g.key == __key) {
          __g.items.push_back(People{p});
          __found = true;
          break;
        }
      }
      if (!__found) {
        __groups.push_back(__struct2{__key, std::vector<People>{People{p}}});
      }
    }
    std::vector<Big> __items;
    for (auto &g : __groups) {
      if (!((((int)g.items.size()) >= 4)))
        continue;
      __items.push_back(Big{g.key, ((int)g.items.size())});
    }
    return __items;
  })();
  (__json(big));
  return 0;
}
